packages/synapse-react-client/src/components/DataGrid/hooks/useGridUndoRedo.tsx

```diff
@@ -0,0 +1,296 @@
+import { DataGridRow, Operation } from '../DataGridTypes'
+import { useCallback, useEffect, useState } from 'react'
+import { ModelChange } from '../utils/applyModelChange'
+import { Button, Menu, MenuItem } from '@mui/material'
+import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown'
+import { convertActionToModelChange } from '../utils/convertActionToModelChange'
+import { useStack } from './useStack'
+
+/**
+ * Represents a single action in the grid that can be undone or redone
+ */
+export type GridAction = {
+  type: 'CREATE' | 'DELETE' | 'UPDATE'
+  rowIndex: number
+  previousValue?: DataGridRow
+  newValue?: DataGridRow
+}
+
+const MAX_UNDO_STEPS = 100
+const MAX_REDO_STEPS = 100
+
+/**
+ * Hook to manage undo/redo functionality for a grid backed by a CRDT model.
+ *
+ * @param onApplyModelChange - Function to apply model changes and commit
+ */
+export function useGridUndoRedo(
+  onApplyModelChange: (change: ModelChange, originalAction: GridAction) => void,
```

Comment:
Seems like this parameter passed to the callback isn't necessary?

```suggestion
  onApplyModelChange: (change: ModelChange) => void,
```